bee-ledger/src/workers/consensus/white_flag.rs

```diff
@@ -68,70 +58,155 @@ fn apply_regular_essence<B: StorageBackend>(
+        transaction_id,
+        essence,
+        inputs.iter().map(|(output_id, input)| (output_id, *input)),
+        unlock_blocks,
+        metadata.milestone_index,
+        metadata.milestone_timestamp,
+    );
+
+    // Validation of the inputs commitment.
+    if context.essence.inputs_commitment() != &context.inputs_commitment {
+        return Ok(ConflictReason::InputsCommitmentsMismatch);
+    }
+
+    // Validation of inputs.
+    for ((output_id, consumed_output), unlock_block) in inputs.iter().zip(unlock_blocks.iter()) {
+        let (conflict, amount, consumed_native_tokens, unlock_conditions) = match consumed_output {
+            Output::Basic(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Alias(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Foundry(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Nft(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            _ => return Err(Error::UnsupportedOutputKind(consumed_output.kind())),
+        };
+
+        if let Err(conflict) = conflict {
+            return Ok(conflict);
+        }
+
+        // TODO maybe put in unlocks methods.
+        if let Some(timelock) = unlock_conditions.timelock() {
+            if *timelock.milestone_index() != 0 && context.milestone_index < timelock.milestone_index() {
+                return Ok(ConflictReason::TimelockMilestoneIndex);
             }
-            Output::Nft(_) => {
-                // TODO
+            if timelock.timestamp() != 0 && context.milestone_timestamp < timelock.timestamp() as u64 {
+                return Ok(ConflictReason::TimelockUnix);
             }
         }
 
-        consumed_outputs.insert(*output_id, consumed_output);
+        context.input_amount = context
+            .input_amount
+            .checked_add(amount)
+            .ok_or(Error::ConsumedAmountOverflow)?;
+        for native_token in consumed_native_tokens.iter() {
+            let native_token_amount = context.input_native_tokens.entry(*native_token.token_id()).or_default();
+            *native_token_amount = native_token_amount
+                .checked_add(*native_token.amount())
+                .ok_or(Error::ConsumedNativeTokensAmountOverflow)?;
+        }
```

Comment:
I would add some newlines here for clarity:
```suggestion
        context.input_amount = context
            .input_amount
            .checked_add(amount)
            .ok_or(Error::ConsumedAmountOverflow)?;

        for native_token in consumed_native_tokens.iter() {
            let native_token_amount = context.input_native_tokens.entry(*native_token.token_id()).or_default();

            *native_token_amount = native_token_amount
                .checked_add(*native_token.amount())
                .ok_or(Error::ConsumedNativeTokensAmountOverflow)?;
        }
```